[metal] Implement a Metal HAL driver

antiagainst opened the pull request:
This pull request resurrects the previously removed Metal
HAL driver due to HAL revamping. It now implements the
HAL driver using Objective-C, instead of Objective-C++
like previously. We also explicitly track resource lifetimes
and manage dependencies, following IREE explicitness.

Right now the following features are implemented:
- HAL driver: device emulation, and some feature printing
- HAL device: resource creation, kernel launches, and sync
- HAL allocation: simple allocator without any caching or sub
  allocation (we expect common layer for those later)
- HAL command buffer: one-shot command buffer, both
  direct and indirect dispatches
- HAL executables: both compile to MSL and `.metallib`
- HAL executable caches: just simple no op cache for now
- HAL builtin executables: for filling/copying buffers
- HAL semaphore: real async queue execute, but not
  stream-ordered allocation yet

The implementation right now assumes Metal3 devices. 
It passes all existing CTS tests, end-to-end StableHLO
and TOSA op and model tests in tree for Apple M1
MacBook.

Fixes https://github.com/openxla/iree/issues/4370

antiagainst committed:
[metal] Add build and registration for a Metal HAL driver

This commit added CMake build configurations for a Metal HAL driver
in the `experimental/` directory. Along the way, added basic logic
for Metal driver registration and device enumeration.

Building this with `-DIREE_EXTERNAL_HAL_DRIVERS=metal`, we can
have `tools/iree-run-module --dump_devices` showing:

```
--device=metal://xxxxxxxxxxxxxxxx
  Apple M1 Pro

- gpu-family: apple7(a14/m1) metal3
- argument-buffer-tier: 2
- simd-scoped-operation: matmul reduce permute
```

antiagainst committed:
[metal] Implement Metal allocator and buffer APIs

Right now the allocator does no advanced caching or suballocating,
which is expected to be supported via an upper layer later. Buffers
are directly created and destroyed as the API calls.

antiagainst committed:
[metal] Wire up creating devices, allocators, and buffers

This commit fills driver APIs for creating devices, and adds a
basic device implementation to create allocator and buffers.
The rest of device APIs are left as unimplemented.

We also start to enable CTS tests since this commit. The first
batch includes driver, allocator, and buffer mapping tests.

antiagainst committed:
[metal] Use Metal shared event to implement IREE semaphore APIs

Metal shared events can be used to synchronize changes to resources
across multiple Metal device objects, across processes, or between
a device object and CPU access to resources. It's a direct match
for IREE semaphores.

antiagainst committed:
[metal] Implement descriptor set and pipeline layout APIs

We don't have direct mappings of these concepts in Metal; so just
capture the content as normal C structs for later usage.

antiagainst committed:
[metal] Support loading executables and compiling kernels

This commit implements IREE executable APIs and adds a no-op
executable cache. This reuses the already in-tree SPIR-V CodeGen
and MSL cross compilation path to generate kernels and embed
them into IREE deployable flatbuffer as MSL source strings.
Then at runtime we compile those source strings into Metal
kernel functions and get pipeline state object out fo them.

A future commit will wire up .mtllib or binary archive offline
compilation.

antiagainst committed:
[metal] Add device parameters to driver/device creation APIs

This commit defines `iree_hal_metal_device_params_t` for
controlling major Metal device behavior. Right now we expose
arena block size and command dispatch type.

antiagainst committed:
[metal] Support one-shot command buffer dispatch and barrier APIs

This commit adds Metal implementation for IREE one-shot command
buffer APIs around lifetime, descriptor set, dispatch, and barriers.
Metal have MTLCommandBuffer command buffer concept for a direct
mapping. Though Metal commands are recorded via another level of
abstraction, command encoders, into command buffers.

antiagainst committed:
[metal] Support push constants in compilers and runtime

Push constants are expected to be directly embedded in the binding
tables and avoid redirections. We use Metal compute command encoder
`setBytes:length:atIndex` for such purposes. This would also require
we teach SPIR-V to MSL cross compilation to emit the proper code
interface for such purposes.

antiagainst committed:
[metal] Implement command buffer fill/copy/update buffer cases

This commit implements command buffer fill/copy buffer APIs for
the cases directly supported by Metal. For update buffer, there
is no direct Metal API for it; so we emulate it with creating a
new buffer and then copy over.

antiagainst committed:
[metal] Implement IREE event APIs with Metal fence

MTLFence is for tracking resource dependencies across command
encoders, which maps to IREE event APIs.

antiagainst committed:
[metal] Delete explicit CTS test lists given all are passing now

antiagainst committed:
[metal] Eanble passing StableHLO/TOSA op end-to-end tests

antiagainst committed:
[metal] Add builtin executables for polyfilling buffer fills

On macOS, the Metal API does not support buffer fills with
non-4-multiple buffer offsets and lengths for fills. It also
does not support 2/4-byte fill patterns. For such cases, we
need to use a kernel dispatch to polyfill it.

Three kernels are provided for different cases, where we can
optimize for 16-byte or 4-byte aligned cases a bit better.

antiagainst committed:
[metal] Use semaphore for implementing queue alloca/dealloca

There is no direct queue ordered allocation support in Metal.
So we just do what other HAL drivers are doing right now--using
semaphore to synchronize for buffer allocation. In the future
there will be an emulation later to replace implementation on
all drivers.

antiagainst committed:
[metal] Add builtin executable for copy unaligned buffers

For buffer copy on macOS, Metal require source/destination offset
and length to be some multiple of 4. So for unaligned cases, we
need a builtin executable to do the copy by ourselves.

This commit adds such a kernel. It copies 1 byte per thread, which
should work for Metal 3 devices. The performance won't likely be
good though; however, this is just a fallback and we can optimize
cases like fill buffer builtin executables if necessary.

antiagainst committed:
[metal] Bump default SPIR-V version to 1.3

SPIR-V 1.3 is what Vulkan 1.1 initially supported. Notably it
adds subgroup operations, which are available on Apple GPUs.
Bumping to this version allow us to generate those subgroup
operations for Metal.

antiagainst committed:
[metal] Add option to allow serial command dispatch for debugging

antiagainst committed:
[metal] Support profiling via GPU frame captures

antiagainst committed:
[metal] Add option to enable resource hazard tracking

IREE is following explicit GPU API model and tracks resource
dependency by itself. So normally we don't need to let Metal
runtime to track resource usages and prevent hazards, which
incurs runtime overhead. But it can be helpful for debugging
purposes.

antiagainst committed:
[metal] Specify dispatch queue QoS class as USER_INITIATED

This is meant to address "Thread Performance Checker: Thread
running at QOS_CLASS_USER_INTERACTIVE waiting on a lower QoS
thread running at QOS_CLASS_DEFAULT." issue on iOS.

antiagainst committed:
[metal] Add option for strong resource reference in command buffers

antiagainst committed:
[metal] Use MTLFence to synchronize encoders in command buffer

This is needed for `MTLHazardTrackingModeTracked` mode per
Metal documentation:

> An MTLFence synchronizes access to one or more resources across
> different passes within a command buffer. Use fences to specify
> any inter-pass resource dependencies within the same command
> buffer.

antiagainst committed:
[metal] Upload initial data for non-shared storage mode buffers

For buffers not in shared storage mode, we need to explicitly
upload the initial data via staging buffers.

antiagainst committed:
[metal] Retain MTLSharedEvent in wait/signal command buffers

IREE would decrease the refcounts of the wrapping semaphores
and release them after `iree_hal_metal_device_queue_execute`
call, but we might still have the wait/signal command buffers
executing on GPU async. This makes sure we don't destroy
resources while in use on GPU.

antiagainst committed:
[metal] Enable compiling to Metal library when possible

This commit extends Metal compilation to additionally invoke
Metal shader compilers on macOS to further compile MSL into
Metal library, so we don't need to JIT compile during runtime.
This reduces runtime overhead. For other platforms where we
don't have access to offline Metal compilers, we still embed
MSL source code in IREE flatbuffer.

antiagainst committed:
[metal] Support different target platforms including iOS

This would enable us to generate Metal libraries for deploying
to iOS and its simulator platforms.

antiagainst committed:
[metal] Enable real async execution on GPU

This commit drops `legacy_sync` path from Metal HAL driver, where
we were forcing waiting all semaphores before all queue execution.
This requires us to track lifetimes of resources involved in queue
execution better, particularly we need to make sure semaphores
and command buffers aren't released until they are not needed by
the GPU.

antiagainst committed:
[metal] Drop IREE HAL event via Metal fence implementation

IREE's HAL event is more flexible than what the Metal fence
supports, so we cannot just directly map there. We'd need
to find better ways to support. Though HAL even is mostly
unused right now so it's fine to be unimplemented.

antiagainst committed:
[metal] Return iree_ok_status for device queue flush

There is nothing to do for now given we immediately release
workload to the GPU on queue execute.

antiagainst committed:
[metal] Use resource set to manage wait/signal semaphores

Resource set would handle resource retaining and releasing on
insert and destroy automatically.

antiagainst committed:
[metal] Use iree_status_t annotation for compute pipeline errors

As a library we shouldn't call into console printing. We already
have scaffolding for rich error reporting via `iree_status_t`
annotations that is more efficient and controllable.

This commit eliminates all `NSLog` calls in the Metal HAL driver.

antiagainst committed:
[metal] Tidy up buffer compatibility and storage mode management

This commit improves the way we decide buffer compatibility from
the Metal allocator:

- Now we switch to request host local + device visible memory for
  the case where device local + host visible is non-existing.
- For cases where we have device local + host visible memory (on
  non-uniform memory macOS), we explicitly manage initial data
  uploading and synchronization for Managed storage mode.

antiagainst committed:
[metal] Implement buffer invalidate range for managed storage

For such storage mode, we need to use `synchronizeResource` in a
blit encoder to make GPU buffers visible to CPU.

antiagainst committed:
[metal] Use MTLEvent for synchronizing when switching encoders

We were using MTLFence objects, but the difference between IREE
HAL and Metal API means we may see many encoder switches. It
would require creating a lot GPU objects. In order to avoid the
cost, we just use one MTLEvent with different values for different
switches.

antiagainst committed:
[metal] Signal and wait MTLEvent for execution only barriers

There is no direct corresponding APIs for execution only barrier
in Metal. We just signal and wait on the same value of a MTLEvent.

antiagainst committed:
[metal] Switch to use command segments for recording command buffer

Unlike Vulkan, Metal adopts a multi-level command recording model--
memory/dispatch commands are not directly recorded into a command buffer;
rather, they must go through the additional level of blit/compute encoders.
IREE's HAL follows the flat Vulkan command buffer recording model, so we
have a mismatch here. Implementing IREE's HAL using Metal would require
switching encoders for interleaved memory and dispatch commands.
Additionally, certain IREE HAL API features do not have direct mapping in
Metal APIs, e.g., various forms of IREE HAL execution/memory barriers.
Translating them would require looking at both previous and next commands
to decide the proper mapping.

Due to these reasons, it's beneficial to have a complete view of the full
command buffer and extra flexibility during recording, in order to fixup
past commands, or inspect future commands.

Therefore, to implement IREE HAL command buffers using Metal, we perform
two steps using a linked list of command segments. First we create segments
(iree_hal_metal_command_buffer_prepare_* and iree_hal_metal_command_segment_create_*)
to keep track of all IREE HAL commands and the associated data, and then,
when finalizing the command buffer, we iterate through all the segments and
record their contents (iree_hal_metal_command_segment_record_*) into a
proper Metal command buffer . A linked list gives us the flexibility to
organize command sequence in low overhead; and a deferred recording gives
us the complete picture of the command buffer when really started recording.

antiagainst committed:
[metal] Drop some unnecessary error checks

These checks should not happen with compilers and runtime validation.
So just turn them into asserts.

antiagainst committed:
[metal] Construct argument buffer at dispatch recording time

This is in preparation of upcoming changes to use one single
device buffer for holding all argument buffers.

antiagainst committed:
[metal] Use staging buffer for argument buffers and update sources

This commit switches the Metal HAL driver to use a staging buffer
for recording argument buffers and uploading buffer update source
data. This avoids creating lots of small-sized buffers like we
did previously, and avoids using command buffer completion callback
to handle their lifetime.

antiagainst committed:
[metal] Add technical details README file

antiagainst committed:
[metal] Add some TODOs for expected changes in command buffer

antiagainst committed:
[metal] NFC: Create struct for state-related command buffer fields

This makes the `iree_hal_metal_command_buffer_t` structure more
clear and readable.

ScottTodd commented on experimental/metal/CMakeLists.txt:
I wonder if we'd want to include a platform check here (`if(APPLE)`?), before adding a library with `-framework Foundation` deps:
https://github.com/openxla/iree/blob/a10566bf657903d7bbc484a17858a806e9a917cc/runtime/src/iree/hal/drivers/CMakeLists.txt#L15-L18

```cmake
if(NOT APPLE)
  message(STATUS "Host platform is not Apple, can't build Metal HAL driver")
  set(IREE_EXTERNAL_METAL_HAL_DRIVER_FOUND FALSE)
endif()
set(IREE_EXTERNAL_METAL_HAL_DRIVER_FOUND TRUE)
```

Maybe not important, considering you already need to set `IREE_EXTERNAL_HAL_DRIVERS` to use this.

antiagainst replied:
Yes this could be useful when moving it out of experimental. I'll leave it as-is for now given this is still in experimental.

ScottTodd commented on experimental/metal/README.md:
Love this README, thanks!

ScottTodd commented on experimental/metal/api.h:
```diff
@@ -0,0 +1,107 @@
+  IREE_HAL_METAL_COMMAND_BUFFER_RESOURCE_REFERENCE_MODE_UNRETAINED = 0,
+  // Letting the command buffer to maintain strong references to used resources.
+  IREE_HAL_METAL_COMMAND_BUFFER_RESOURCE_REFERENCE_MODE_RETAINED = 1,
+} iree_hal_metal_command_buffer_resource_reference_mode_t;
```
grammar nits

```suggestion
typedef enum iree_hal_metal_command_buffer_resource_reference_mode_e {
  // Do not maintain strong references to resources used in command buffers.
  IREE_HAL_METAL_COMMAND_BUFFER_RESOURCE_REFERENCE_MODE_UNRETAINED = 0,
  // Maintain strong references to resources used in command buffers.
  IREE_HAL_METAL_COMMAND_BUFFER_RESOURCE_REFERENCE_MODE_RETAINED = 1,
} iree_hal_metal_command_buffer_resource_reference_mode_t;
```

ScottTodd commented on experimental/metal/api.h:
```diff
@@ -0,0 +1,107 @@
+  // Letting the Metal runtime to prevent hazards when modifying this object's
+  // contents.
+  IREE_HAL_METAL_RESOURCE_HAZARD_TRACKING_MODE_TRACKED = 1,
+} iree_hal_metal_resource_hazard_tracking_mode_t;
```
I'm not sure I follow here. Is "Metal runtime" something we're using, or is it the IREE Metal HAL driver itself?

```c
  // Resource hazard tracking mode.
  // IREE is following explicit GPU API model and tracks resource dependency by
  // itself. So normally we don't need to let Metal runtime to track resource
  // usages and prevent hazards, which incurs runtime overhead. But it can be
  // helpful for debugging purposes.
  iree_hal_metal_resource_hazard_tracking_mode_t resource_hazard_tracking_mode;
```

```suggestion
typedef enum iree_hal_metal_resource_hazard_tracking_mode_e {
  // Do not track resource hazards. Hosting applications are responsible
  // for ensuring that resources are not modified while in use.
  IREE_HAL_METAL_RESOURCE_HAZARD_TRACKING_MODE_UNTRACKED = 0,
  // Track resource hazards within the Metal runtime.
  IREE_HAL_METAL_RESOURCE_HAZARD_TRACKING_MODE_TRACKED = 1,
} iree_hal_metal_resource_hazard_tracking_mode_t;
```

antiagainst replied:
"Metal runtime" means the underlying Metal runtime in the system; not IREE Metal driver. I added "system" to prefix to be clear.

ScottTodd commented on compiler/src/iree/compiler/Dialect/HAL/Target/MetalSPIRV/MSLToMetalLib.cpp:
```diff
@@ -0,0 +1,95 @@
+      sdk = "iphonesimulator";
+      break;
+  }
+  return llvm::Twine("xcrun -sdk ")
+      .concat(sdk)
+      .concat(" metal -c ")
+      .concat(mslFile)
+      .concat(" -o - | xcrun -sdk ")
+      .concat(sdk)
+      .concat(" metallib - -o ")
+      .concat(libFile)
+      .str();
```
This could use a comment (why pipe?)

ScottTodd commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,522 @@
+    device->capture_manager = NULL;
+    *out_device = (iree_hal_device_t*)device;
+  }
+  return iree_ok_status();
```
Return `status`, not `iree_ok_status()` (also maybe audit the rest of the code for that error)
```suggestion
  return status;
```

antiagainst replied:
Good catch!

ScottTodd commented on experimental/metal/metal_driver.m:
```diff
@@ -0,0 +1,450 @@
+  return 0;
+}
+
+static const char* iree_hal_metal_get_gpu_family_name(MTLGPUFamily family) {
+  switch (family) {
+    case MTLGPUFamilyApple8:
+      return "apple8(a15/m2)";
+    case MTLGPUFamilyApple7:
+      return "apple7(a14/m1)";
+    case MTLGPUFamilyApple6:
+      return "apple6(a13)";
+    case MTLGPUFamilyApple5:
+      return "apple5(a12)";
+    case MTLGPUFamilyApple4:
+      return "apple4(a11)";
+    case MTLGPUFamilyApple3:
+      return "apple3(a9/a10)";
+    case MTLGPUFamilyApple2:
+      return "apple2(a8)";
+    case MTLGPUFamilyApple1:
+      return "apple1(a7)";
```
Do we need all of these? Seems like some binary size bloat.

antiagainst replied:
I'd prefer to keep for the usefulness--just a few bytes more anyway. :)

ScottTodd commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1187 @@
+  if (iree_status_is_ok(status)) {
+    iree_hal_command_buffer_initialize(
+        device, mode, command_categories, IREE_HAL_QUEUE_AFFINITY_ANY, binding_capacity,
+        &iree_hal_metal_command_buffer_vtable, &command_buffer->base);
```
I think there's a logic error in here if the malloc fails. The code after the if expects `command_buffer->base` to be set:
```c
  *out_command_buffer = &command_buffer->base;
  IREE_TRACE_ZONE_END(z0);
  return status;
```

antiagainst replied:
Good catch. Moved to the previous `if` case.

ScottTodd commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1187 @@
+      iree_hal_resource_set_insert(command_buffer->resource_set, 1, &target_buffer));
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
```
Change to 
```c
if (iree_status_is_ok(status)) {
  status = ...
}
```
style, so the trace zone is not left open if there is an error

```suggestion
  if (iree_status_is_ok(status)) {
    status = iree_hal_resource_set_insert(command_buffer->resource_set, 1, &target_buffer));
  }      

  IREE_TRACE_ZONE_END(z0);
  return status;
```

(or use `IREE_RETURN_AND_END_ZONE_IF_ERROR`)

antiagainst replied:
Good point!

ScottTodd commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1187 @@
+                                               /*alignment=*/4, &offset));
+
+  IREE_RETURN_IF_ERROR(
+      iree_hal_resource_set_insert(command_buffer->resource_set, 1, &target_buffer));
```
```suggestion
  IREE_RETURN_AND_END_ZONE_IF_ERROR(
      z0, iree_hal_resource_set_insert(command_buffer->resource_set, 1, &target_buffer));
```

ScottTodd commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1187 @@
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  const iree_hal_buffer_t* buffers[2] = {source_buffer, target_buffer};
+  IREE_RETURN_IF_ERROR(iree_hal_resource_set_insert(command_buffer->resource_set, 2, buffers));
```
```suggestion
  IREE_RETURN_AND_END_ZONE_IF_ERROR(z0, iree_hal_resource_set_insert(command_buffer->resource_set, 2, buffers));
```

ScottTodd commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1187 @@
+  IREE_ASSERT(set != IREE_HAL_METAL_PUSH_CONSTANT_BUFFER_INDEX);
+  iree_hal_metal_descriptor_t* descriptors = command_buffer->state.current_descriptors;
+  IREE_ASSERT(iree_hal_metal_is_sorted_unique_descriptors(
+      descriptors, command_buffer->state.current_total_binding_count));
```
Asserts intended? Are these really never possible, or should this return an error instead?

antiagainst replied:
Yes. Though this does not matter anymore--I changed how those descriptors are handled so no need to sort anymore.

ScottTodd reviewed (commented):
Reviewed mostly for structure. Not familiar enough with the API to review correctness / architecture.

antiagainst committed:
[metal] Use separate lists for different descriptor sets

This simplifies the logic for descriptor tracking in command
buffer states. Now we don't need to find the proper starting
index and sort.

benvanik commented on compiler/src/iree/compiler/Dialect/HAL/Target/MetalSPIRV/MetalSPIRVTarget.cpp:
```diff
@@ -145,41 +161,59 @@ class MetalSPIRVTargetBackend : public TargetBackend {
+      // We need to use offline Metal shader compilers. They are only available
+      // on macOS devices.
+      auto hostTriple = llvm::Triple(llvm::sys::getProcessTriple());
+      if (hostTriple.isMacOSX()) {
```
would be nice to fix this logic so that it instead tested for the presence of the tool instead - the metal tools are available for windows (and hopefully more platforms in the future!) - if you file an issue for windows support for metallib and tag this with a TODO(#) I can take a look

antiagainst replied:
https://github.com/openxla/iree/issues/14048 for tracking this one.

benvanik replied:
thanks for filing!

benvanik commented on compiler/src/iree/compiler/Dialect/HAL/Target/MetalSPIRV/MetalSPIRVTarget.cpp:
```diff
@@ -145,41 +161,59 @@ class MetalSPIRVTargetBackend : public TargetBackend {
+      // on macOS devices.
+      auto hostTriple = llvm::Triple(llvm::sys::getProcessTriple());
+      if (hostTriple.isMacOSX()) {
+        for (auto [shader, entryPoint] :
```
we should talk about linking strategies - we really need to be shipping a single metallib and instantiating a single hal executable; I can understand how the SPIR-V lowering makes that tricky but it could be something we conditionally enable when compiling to metallibs. compileMSLToMetalLib would change to a two phase compile each to AIR then link all AIR to metallib. this will shrink binary size, improve startup performance, dramatically simplify the VMFB (instead of managing hundreds of executables we'd manage 1), etc etc. since it's performance instead of correctness you can file an issue for metal linking and tag with a TODO(#).

antiagainst replied:
Yup. Created https://github.com/openxla/iree/issues/14047 for this.

benvanik replied:
thanks for filing!

benvanik commented on experimental/metal/README.md:
```diff
@@ -0,0 +1,351 @@
+# IREE Metal HAL Driver
+
+This document lists technical details regarding the Metal HAL driver. Note that
+the Metal HAL driver is working in progress and experimental; this document is
```
```suggestion
the Metal HAL driver is a work in progress and experimental; this document is
```

benvanik commented on experimental/metal/README.md:
```diff
@@ -0,0 +1,351 @@
+
+#### Collectives
+
+Collectives suppport is unimplemented yet.
```
```suggestion
Collectives suppport is not yet implemented.
```

benvanik commented on experimental/metal/README.md:
```diff
@@ -0,0 +1,351 @@
+
+### Executable cache
+
+IREE [`iree_hal_executable_cache_t`][hal-executable-cache] is modelling a cache of
```
```suggestion
IREE [`iree_hal_executable_cache_t`][hal-executable-cache] is modeling a cache of
```

benvanik commented on experimental/metal/README.md:
```diff
@@ -0,0 +1,351 @@
+
+IREE [`iree_hal_executable_cache_t`][hal-executable-cache] is modelling a cache of
+preprared GPU executables for a particular device. At the moment the Metal
+HAL driver does not peforming any cache on GPU programs; it simply reads the
```
```suggestion
HAL driver does not peforming any caching of GPU programs; it simply reads the
```

benvanik commented on experimental/metal/builtin_executables.m:
```diff
@@ -0,0 +1,212 @@
+  uint64_t buffer_offset;  // Buffer offset to fill (in bytes)
+  uint64_t buffer_length;  // Buffer length to fill (in bytes)
+  uint32_t pattern;        // 32-bit fill pattern
+} iree_hal_metal_buffer_fill_spec;
```
```suggestion
} iree_hal_metal_buffer_fill_spec_t;
```
(here and on others)

benvanik commented on experimental/metal/builtin_executables.m:
```diff
@@ -0,0 +1,212 @@
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_metal_builtin_executable_t* executable = NULL;
+  unsigned entry_point_count = IREE_ARRAYSIZE(iree_hal_metal_builtin_executable_entry_points);
```
```suggestion
  iree_host_size_t entry_point_count = IREE_ARRAYSIZE(iree_hal_metal_builtin_executable_entry_points);
```

benvanik commented on experimental/metal/builtin_executables.m:
```diff
@@ -0,0 +1,212 @@
+    MTLCompileOptions* compile_options = [MTLCompileOptions new];  // +1
+    compile_options.languageVersion = MTLLanguageVersion3_0;
+
+    for (unsigned i = 0; i < IREE_ARRAYSIZE(iree_hal_metal_builtin_executable_entry_points); ++i) {
```
```suggestion
    for (iree_host_size_t i = 0; i < IREE_ARRAYSIZE(iree_hal_metal_builtin_executable_entry_points); ++i) {
```
(worth find/replacing `unsigned` with `iree_host_size_t` if you do this elsewhere)

benvanik commented on experimental/metal/builtin_executables.m:
```diff
@@ -0,0 +1,212 @@
+  if (iree_status_is_ok(status)) {
+    *out_executable = executable;
+  } else {
+    iree_allocator_free(host_allocator, executable);
```
you want to call iree_hal_metal_builtin_executable_destroy here instead so that if there's partially populated values they are released properly (will need to do this in other places too if you've used this pattern)

benvanik commented on experimental/metal/builtin_executables.m:
```diff
@@ -0,0 +1,212 @@
+  }
+
+  IREE_TRACE_ZONE_END(z0);
+  return iree_ok_status();
```
```suggestion
  return status;
```

benvanik commented on experimental/metal/builtin_executables.m:
```diff
@@ -0,0 +1,212 @@
+    uint32_t pattern) {
+  id<MTLComputePipelineState> pso = nil;
+  MTLResourceUsage usage = MTLResourceUsageWrite;
+  const unsigned workgroup_size = 32;
```
(for the entire PR) you should almost never use `unsigned` - this should be `iree_device_size_t` - each of these `unsigned` is one more thing limiting us to 32-bit only even if metal started supporting larger sizes. it's also not good to have inconsistent type widths (if length is 64-bit then it's implicitly cast to unsigned here for the ceil div and that kind of stuff is really hard to track down).

antiagainst replied:
Makes sense. Did a thorough pass to audit usages of `unsigned`, `int`, `int32_t`, and `uint32_t`.

benvanik commented on experimental/metal/direct_allocator.m:
```diff
@@ -0,0 +1,363 @@
+  if (iree_status_is_ok(status)) {
+    iree_hal_resource_initialize(&iree_hal_metal_allocator_vtable, &allocator->resource);
+    allocator->base_device = base_device;
+    iree_hal_device_retain(base_device);
```
don't retain the HAL device here as that creates a cycle (the device owns the allocator and here the allocator is owning the device)

benvanik commented on experimental/metal/metal_buffer.m:
```diff
@@ -0,0 +1,199 @@
+  if (buffer->buffer.storageMode == MTLStorageModeManaged) {
+    const iree_hal_device_t* device =
+        iree_hal_metal_allocator_device(buffer->base.device_allocator);
+    id<MTLCommandQueue> queue = iree_hal_metal_device_command_queue(device);
```
this is not safe: buffer->base.device_allocator may be _any_ allocator (caching allocator, suballocator, etc) and can't be reinterpret-cast to a metal allocator. if this is required (unfortunate!) then you'll likely want to store the MTLCommandQueue on the buffer #if IREE_PLATFORM_MACOS (buffer create could always take it as an arg but only retain it in the struct if on macos)

antiagainst replied:
Yeah, AFAICT, it's necessary to refer to the command queue here to be able issue commands. Good suggestions! Done.

benvanik commented on experimental/metal/direct_allocator.m:
```diff
@@ -0,0 +1,363 @@
+  return allocator->host_allocator;
+}
+
+const iree_hal_device_t* iree_hal_metal_allocator_device(
```
this is unsafe and should be removed (see comment on the buffer invalidate) - you can then also stop taking the iree_hal_device_t on the allocator and just use the MTLDevice (good decoupling)

benvanik commented on experimental/metal/direct_allocator.m:
```diff
@@ -0,0 +1,363 @@
+    iree_hal_buffer_t* IREE_RESTRICT base_buffer) {
+  iree_hal_metal_allocator_t* allocator = iree_hal_metal_allocator_cast(base_allocator);
+
+  IREE_TRACE_FREE_NAMED(IREE_HAL_METAL_ALLOCATOR_ID, (void*)metal_buffer);
```
metal_buffer is undefined (need to get the iree_hal_metal_buffer_device_pointer) - may be other tracing errors hiding so worth a build with it enabled

benvanik commented on experimental/metal/direct_allocator.m:
```diff
@@ -0,0 +1,363 @@
+      options = MTLResourceStorageModeManaged;
+#else
+      options = MTLResourceStorageModeShared;
+#endif
```
```suggestion
#endif  // IREE_PLATFORM_MACOS
```

benvanik commented on experimental/metal/direct_allocator.m:
```diff
@@ -0,0 +1,363 @@
+        metal_buffer, base_allocator, compat_params.type, compat_params.access, compat_params.usage,
+        allocation_size, /*byte_offset=*/0,
+        /*byte_length=*/allocation_size, iree_hal_buffer_release_callback_null(), &buffer);  // +1
+  }
```
can just return the RESOURCE_EXHAUSTED error instead of carrying the status down and then initialize the status from the iree_hal_metal_buffer_wrap call instead

```suggestion
  bool has_init_data = !iree_const_byte_span_is_empty(initial_data);

  MTLResourceOptions options =
      iree_hal_metal_select_resource_options(compat_params.type, allocator->is_unified_memory,
                                             has_init_data, allocator->resource_tracking_mode);
  id<MTLBuffer> metal_buffer = nil;
  // If we chose shared storage mode, we can handle initial data with newByfferWithBytes directly.
  // Otherwise we just create the buffer here, and explicitly transfer range later.
  if (has_init_data && iree_all_bits_set(options, MTLResourceStorageModeShared)) {
    IREE_ASSERT_EQ(allocation_size, initial_data.data_length);
    metal_buffer = [allocator->device newBufferWithBytes:(void*)initial_data.data
                                                  length:initial_data.data_length
                                                 options:options];  // +1
  } else {
    metal_buffer = [allocator->device newBufferWithLength:allocation_size options:options];  // +1
  }

  IREE_TRACE_ZONE_END(z0);
  if (!metal_buffer) {
    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED, "unable to allocate buffer");
  }

  iree_hal_buffer_t* buffer = NULL;
  iree_status_t status = iree_hal_metal_buffer_wrap(
        metal_buffer, base_allocator, compat_params.type, compat_params.access, compat_params.usage,
        allocation_size, /*byte_offset=*/0,
        /*byte_length=*/allocation_size, iree_hal_buffer_release_callback_null(), &buffer);  // +1
```

benvanik commented on experimental/metal/direct_allocator.m:
```diff
@@ -0,0 +1,363 @@
+  }
+
+  if (iree_status_is_ok(status) && has_init_data) {
+    if (iree_all_bits_set(options, MTLResourceStorageModePrivate)) {
```
you can validate this above to avoid creating the buffer as has_init_data and options are both known earlier

benvanik commented on experimental/metal/direct_command_buffer.h:
```diff
@@ -0,0 +1,65 @@
+// are not directly supported by Metal APIs.
+//
+// |out_command_buffer| must be released by the caller (see
+// |iree_hal_command_buffer_release|).
```
```suggestion
// iree_hal_command_buffer_release).
```

benvanik commented on experimental/metal/direct_command_buffer.h:
```diff
@@ -0,0 +1,65 @@
+// The command buffer would have the given |mode| and be recorded and submitted
+// to the given |queue|.
+//
+// If |block_pool| is non-NULL then the command buffer will retain copies of
```
I believe the block pool is required

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_metal_command_buffer_t* command_buffer = NULL;
+  iree_status_t status =
```
could IREE_RETURN_AND_END_ZONE_IF_ERROR on this (since nothing needs to be cleaned up yet)

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+  // Allocate the command segment and keep track of all necessary API data.
+  uint8_t* storage_base = NULL;
+  iree_hal_metal_command_segment_t* segment = NULL;
+  iree_status_t status = iree_arena_allocate(
```
could IREE_RETURN_AND_END_ZONE_IF_ERROR here (and on the others) as there's nothing to clean up (removes a lot of the if(ok) nesting and allows for unconditional returns of iree_status_ok to help the compiler)

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+  // Note that fillBuffer:range:value: only accepts a single byte as the pattern but FillBuffer
+  // can accept 1/2/4 bytes. If the pattern itself contains repeated bytes, we can call into
+  // fillBuffer:range:value:. Otherwise we need to emulate the support.
+  uint8_t single_byte_value = 0u;
```
making iree_hal_metal_get_duplicated_single_byte_value return the uint32_t (as there's never a case where it can fail) will make this function a bit easier to read (can_use_metal_api becomes just the check above and not also this & and then the if below where this function is called again)

antiagainst replied:
Not sure I follow exactly here. `fillBuffer:range:value:` requires a `uint8_t` as the fill value so we need to get the `uint8_t` out of it; cannot do with `uint32_t`. Do you mean merging the logic of `iree_hal_metal_get_duplicated_single_byte_value` and `iree_hal_metal_duplicate_to_four_byte_value` together so we have one single function for them? I'd prefer to keep them separate for better focus. 

I tried to re-arrange the logic a bit to make it less branchy though.

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+  return status;
+}
+
+static iree_status_t iree_hal_metal_command_buffer_prepare_copy_buffer(
```
nit: ordering here makes it a bit tricky to read - can you put prepare_copy_buffer > record_copy_buffer > prepare_update_buffer > record_update_buffer?

antiagainst replied:
There is no record_update_buffer as it's reusing the copy buffer segment there. So maybe that's why it's a bit weird. :)

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+  }
+
+  memcpy((uint8_t*)&command_buffer->state.push_constants + offset, values, values_length);
+  command_buffer->state.current_pipeline_layout = pipeline_layout;
```
instead of stashing the layout you can stash whatever you need from it - in this case the push_constant count from iree_hal_metal_pipeline_layout_push_constant_count - to avoid retaining objects (if you did cache the layout here you'd need to retain it because the caller is allowed to release it immediately after calling this function)

**actually...**

thinking about it, it's even trickier - I think you'll need to drop the current_pipeline_layout entirely _and_ move the push constant count to each entry point - it's legal to do:
```
push_constants(pipeline_with_2_constants, ...)
dispatch()
push_descriptor_set(pipeline_with_1_constant, ...)
dispatch()
```
and other such things - basically, how many push constants are used by a dispatch is specific to the entry point used in that dispatch and not a property of the command buffer state

you already have the layout in iree_hal_metal_kernel_params_t which you get in the dispatch, so you can just ask it for the push constant count in prepare_dispatch directly and not track any such state on the command buffer!

antiagainst replied:
Ah, exactly! Thanks for the suggestion! :D

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+  iree_host_size_t descriptor_count = 0;
+  // Calculate the total number of bindings across all descriptor sets.
+  for (int i = 0; i < IREE_HAL_METAL_PUSH_CONSTANT_BUFFER_INDEX; ++i) {
+    descriptor_count += command_buffer->state.descriptor_sets[i].active_count;
```
this is something you could store on the pipeline layout in order to avoid needing this logic (similar to push_constant count) - I suspect there are issues with tracking this state just like with the push constants

a fixed-size array of 4 uint8_t's on the kernel_params or pipeline layout would probably do the trick

antiagainst replied:
Actually we already have such information in the pipeline layout. So switched to use that. :)

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+    // Build argument encoder and argument buffer for the current descriptor set.
+    // TODO(antiagainst): Use a cache layer to cache and reuse argument buffers with the same
+    // content, to avoid duplicating overhead.
+    id<MTLBuffer> argument_buffer = command_buffer->staging_buffer->metal_buffer;
```
maybe this logic could be done during preparation?

antiagainst replied:
It's quite coupled with argument encoder and compute encoder; so not sure how much benefits it brings. I'll leave it as-is for now.

benvanik reviewed (changes requested):
(still working my way through but have to go afk for a bit - about halfway through command_buffer.m!)

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+       segment = segment->next_segment) {
+    switch (segment->action) {
+      case IREE_HAL_METAL_COMMAND_SEGMENT_ACTION_BARRIER:
+        iree_hal_metal_command_segment_record_barrier(command_buffer, &segment->barrier);
```
all of these need to return on error
(I'm seeing quite a few places missing returns - would be worth doing a full scrub to ensure iree_status_t is never dropped on the floor)

antiagainst replied:
Ouch! Great catch! Totally missed this.. Did a thorough pass over to make sure they are propagated.

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+    iree_hal_command_buffer_t* base_command_buffer) {
+  iree_hal_metal_command_buffer_t* command_buffer =
+      iree_hal_metal_command_buffer_cast(base_command_buffer);
+  iree_hal_metal_command_segment_record(command_buffer);
```
this needs to propagate the error

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+  // Record the dispatch, either direct or indirect.
+  uint32_t* workgroup_size = segment->kernel_params.threadgroup_size;
+  if (segment->workgroups_buffer == nil) {  // Direct dispatch
+    uint32_t* workgroup_count = segment->workgroup_count;
```
style nit: put the comments one line down instead of after the if - it's harder to scan and spot them as they're in random horizontal positions
```suggestion
  if (segment->workgroups_buffer == nil) {
    // Direct dispatch of a fixed workgroup count.
    uint32_t* workgroup_count = segment->workgroup_count;
```

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -0,0 +1,1153 @@
+
+  // Record argument buffers for all descriptors and record buffer usages.
+  iree_hal_metal_descriptor_t* descriptors = segment->descriptors;
+  for (int i = 0; i < segment->descriptor_count;) {
```
iree_host_size_t to match descriptor_count (always keep types consistent)
probably elsewhere in here too

benvanik commented on experimental/metal/metal_buffer.m:
```diff
@@ -0,0 +1,199 @@
+  IREE_TRACE_ZONE_BEGIN(z0);
+  IREE_TRACE_ZONE_APPEND_VALUE(z0, (int64_t)iree_hal_buffer_allocation_size(base_buffer));
+
+  [buffer->buffer release];  // -1
```
you want this after the release callback (I believe) - otherwise the callback would explode if it actually tried to get the MTLBuffer

benvanik commented on experimental/metal/metal_buffer.m:
```diff
@@ -0,0 +1,199 @@
+
+  iree_status_t status = iree_ok_status();
+#if defined(IREE_PLATFORM_MACOS)
+  if (iree_hal_metal_require_autosync_cpu_gpu(base_buffer, buffer->buffer)) {
```
only needed if mapping for READ (any_bit_set(..., MEMORY_ACCESS_READ))

benvanik commented on experimental/metal/metal_device.h:
```diff
@@ -0,0 +1,42 @@
+// Creates a Metal device by wrapping |device| from the given |driver| with the
+// specific |params|.
+//
+// |out_device| must be released by the caller (see |iree_hal_device_release|).
```
```suggestion
// |out_device| must be released by the caller (see iree_hal_device_release).
```

benvanik commented on experimental/metal/metal_device.h:
```diff
@@ -0,0 +1,42 @@
+
+// Returns the Metal command queue associated with the given |base_device|.
+// Note that right now we only support one command queue per Metal device.
+id<MTLCommandQueue> iree_hal_metal_device_command_queue(
```
would be good not to use this (will make it hard to support more queues/devices in the future - this has bit the CUDA side which made this same kind of assumption)
I think you can remove this call after fixing the comment in iree_hal_metal_buffer_invalidate_range

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+  // For polyfilling fill/copy/update buffers that are not directly supported by Metal APIs.
+  iree_hal_metal_builtin_executable_t* builtin_executable;
+
+  // A dispatch queue and associated event listener for running Objective-C blocks to singal
```
```suggestion
  // A dispatch queue and associated event listener for running Objective-C blocks to signal
```

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+// Copyright 2022 The IREE Authors
```
```suggestion
// Copyright 2023 The IREE Authors
```

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+  iree_hal_metal_staging_buffer_t staging_buffer;
+
+  // Original driver that owns this device.
+  iree_hal_driver_t* driver;
```
remove this and instead just pass in the params - drivers are optional and shouldn't be required by the device (this will let us wrap existing user MTLDevices and such) as they are just a factory for creating devices.

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+        metal_device, params->queue_uniform_buffer_size, &device->staging_buffer);
+  } else {
+    iree_hal_device_release((iree_hal_device_t*)device);
+  }
```
this can be refactored to have a more canonical flow (allocate, initialize, then do the subresources) - fewer exits that need to do the device release that way

as it is it's not valid - until you call iree_hal_resource_initialize you can't call iree_hal_device_release - these *must* come after you initialize the device fully

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+                                                          iree_hal_queue_affinity_t queue_affinity,
+                                                          iree_hal_channel_params_t params,
+                                                          iree_hal_channel_t** out_channel) {
+  return iree_make_status(IREE_STATUS_UNIMPLEMENTED, "unimplmented channel create");
```
```suggestion
  return iree_make_status(IREE_STATUS_UNIMPLEMENTED, "collectives not yet implemented");
```

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+      device->command_buffer_resource_reference_mode, device->queue, device->host_allocator,
+      &device->block_pool, &device->staging_buffer, device->builtin_executable, out_command_buffer);
+  if (iree_status_is_ok(status)) {
+    iree_hal_metal_staging_buffer_increase_refcount(&device->staging_buffer);
```
this should be in the command_buffer_create (same place the decrement should be)
I see you have the decrement in submission now but that's not safe - it's possible to create command buffers that are never submitted and this would cause a leak. Since the command buffer lifetime must extend past its execution you can just put the retain/release in the command buffer and manage it all together.

antiagainst replied:
Good catch!

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+
+static iree_status_t iree_hal_metal_device_create_event(iree_hal_device_t* base_device,
+                                                        iree_hal_event_t** out_event) {
+  return iree_make_status(IREE_STATUS_UNIMPLEMENTED, "unimplemented fence create");
```
```suggestion
  return iree_make_status(IREE_STATUS_UNIMPLEMENTED, "unimplemented event create");
```

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+  iree_hal_metal_device_t* device = iree_hal_metal_device_cast(base_device);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  @autoreleasepool {
```
this can be optimized such that you have a single resource set that you insert all of the the wait, signal, and command buffers into and then one completed callback that you free the set

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+    }
+
+    // Finally create a new command buffer and encode signal commands for all signal semaphores.
+    if (signal_semaphore_list.count > 0) {
```
if there's >0 command_buffer_count then you could encoding the signal event directly on the last command buffer submitted and avoid needing to do this additional submission. you'll still need to handle the case where count==0, though.

this way in the steady state we have 0 to 1 command buffers (depending on whether there are waits) and 1 device->host callback. in the future when we do basic wait elision (if the last signaler of the event was the queue being submitted to then no need to wait) then we'll go down to basically commit + callback to cleanup.

```
iree_hal_resource_set_t* submission_resource_set;
if (wait_count) {
  new command buffer
  insert wait semaphores into the set
  commit
}
id<MTLCommandBuffer> last_command_buffer;
if (command_buffer_count > 0) {
  for ... {
    insert command buffer in resource set
    if (!last) commit
  }
  last_command_buffer = command_buffers[count-1];
} else {
  last_command_buffer = new command buffer
}
if (signal_count > 0) {
  [last_command_buffer encodeSignalEvent ...];
  insert signal semaphores in resource set
}
[last_command_buffer addCompletedHandler: free shared resource set];
[last_command_buffer commit];
```

antiagainst replied:
Nice!

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+      iree_hal_resource_set_t* resource_set;
+      IREE_RETURN_AND_END_ZONE_IF_ERROR(
+          z0, iree_hal_resource_set_allocate(&device->block_pool, &resource_set));
+      iree_hal_resource_set_insert(resource_set, signal_semaphore_list.count,
```
these can fail - ensure you're checking iree_status_t results here and elsewhere (I suspect the file needs a scrub)

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -0,0 +1,521 @@
+      iree_hal_resource_set_insert(resource_set, signal_semaphore_list.count,
+                                   signal_semaphore_list.semaphores);
+
+      MTLCommandBufferDescriptor* descriptor = [MTLCommandBufferDescriptor new];  // +1
```
you can cache this descriptor on the device so you don't need to make a new one all the time

antiagainst replied:
Good point!

benvanik commented on experimental/metal/metal_driver.m:
```diff
@@ -0,0 +1,450 @@
+
+  // Try parsing as a device ID.
+  uint8_t device_registry_id[8] = {0};
+  if (iree_string_view_parse_hex_bytes(device_path, 8, device_registry_id)) {
```
```suggestion
  if (iree_string_view_parse_hex_bytes(device_path, IREE_ARRAYSIZE(device_registry_id), device_registry_id)) {
```

benvanik commented on experimental/metal/metal_kernel_library.h:
```diff
@@ -0,0 +1,64 @@
+// Metal represents compute kernels as MTLFunctions. MTLLibrary is just an
+// allocation of MTLFunctions. One creates a MTLComputePipelineState from a
+// MTLFunction and uses the pipeline state for creating compute pipelines.
+// This class bundles all the necesary Metal objects for getting pipeline state
```
```suggestion
// This class bundles all the necessary Metal objects for getting pipeline state
```
(I recommend turning on spell checker in your IDE of choice)

benvanik commented on experimental/metal/metal_kernel_library.h:
```diff
@@ -0,0 +1,64 @@
+// objects for a compute kernel.
+//
+// |out_executable| must be released by the caller (see
+// |iree_hal_executable_release|).
```
```suggestion
// iree_hal_executable_release).
```

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+}
+
+// Verifies the structure of the flatbuffer so that we can avoid doing so during runtime.
+
```
```suggestion
//
```

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+  if (entry_point_count != threadgroup_size_count) {
+    return iree_make_status(
+        IREE_STATUS_INVALID_ARGUMENT,
+        "the numbers of entry points (%zu) and thread group sizes (%zu) mismatch",
```
```suggestion
        "entry points (%zu) and thread group sizes (%zu) count mismatch",
```

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+    if (entry_point_count != shader_library_count) {
+      return iree_make_status(
+          IREE_STATUS_INVALID_ARGUMENT,
+          "the numbers of entry points (%zu) and source libraries (%zu) mismatch",
```
```suggestion
          "entry points (%zu) and source libraries (%zu) count mismatch",
```

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+
+    if (entry_point_count != shader_source_count) {
+      return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                              "the numbers of entry points (%zu) and source strings (%zu) mismatch",
```
```suggestion
                              "entry points (%zu) and source strings (%zu) count mismatch",
```

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+  flatbuffers_string_vec_t shader_libraries_vec =
+      iree_hal_metal_ExecutableDef_shader_libraries_get(executable_def);
+  size_t shader_library_count = flatbuffers_string_vec_len(shader_libraries_vec);
+  if (shader_library_count) {
```
you can unnest these as the for's/if's are fine if the counts are 0 - keeps the logic a bit easier to follow/reduces nesting

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+  return iree_status_annotate_f(status, "for entry point '%s' in MTLLibrary\n", entry_point);
+}
+
+iree_status_t iree_hal_metal_compile_msl(const char* source_code, const char* entry_point,
```
out_* prefix on output arguments (here and in the header)

benvanik replied:
```suggestion
iree_status_t iree_hal_metal_compile_msl(iree_string_view_t source_code, const char* entry_point,
```
you have the length in the caller and can pass that in to NSString initWithBytes:length:encoding: to avoid making NSString scan for the NUL terminator

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+  return iree_ok_status();
+}
+
+iree_status_t iree_hal_metal_load_mtllib(const char* source_lib, size_t length,
```
out_ prefixes

benvanik replied:
```suggestion
iree_status_t iree_hal_metal_load_mtllib(iree_const_byte_span_t source_data,
```
(const char* implies NUL terminated strings, which this is not)

benvanik replied:
we should change this into a two stage process - MTLLibrary loading and then entry_point+MTLFunction+MTLComputePipelineState creation. this would also allow you to share more code between these two paths - today the only difference is whether you call newLibraryWithSource or newLibraryWithData and if you hoisted that out the rest would be the same. it'll also be needed when we (hopefully soon) are putting multiple entry points in a single library.

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+
+    // Try to load as Metal library first. Otherwise, compile each MSL source string into a
+    // MTLLibrary and get the MTLFunction for the entry point to build the pipeline state object.
+    // TODO(antiagainst): We are performing synchronous compilation at runtime here. This is good
```
splitting MTLLibrary from pipeline creation would be good - we should be able to make the MTLLibrary creation async (set a completion handler) and then fork out to newComputePipelineStateWithFunction:completionHandler: for each function in the library so that we get everything parallelized. actually doing it async can come later but making sure we aren't baking in the assumption that there's one library per function and that they are all created synchronously in a loop would be nice.

antiagainst replied:
Totally. It's a bit unfortunate that this is the case right now; but I agree with you regarding where this should be going. Created https://github.com/openxla/iree/issues/14047 for this.

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+    MTLCompileOptions* compile_options = [MTLCompileOptions new];  // +1
+    compile_options.languageVersion = MTLLanguageVersion3_0;
+
+    for (size_t i = 0, e = iree_max(shader_library_count, shader_source_count); i < e; ++i) {
```
this is tricky to read/ensure it works as it's assuming globally true state (only the shader library count or shader source count can be set to non-zero) and sinking the decision of how to load things down below. we should really fix the flatbuffer such that there are libraries and then entry points instead of a bunch of flat lists - an EntryPointDef that had the name, library index to use in the libraries table, and threadgroup info (among whatever else we want to add) would by construction prevent the kind of inconsistencies that arise around this stuff.

definitely feel like the next step after this lands will be to fix linking/the flatbuffer/this creation logic.

antiagainst replied:
Yup. Created https://github.com/openxla/iree/issues/14047 for this.

benvanik commented on experimental/metal/metal_kernel_library.m:
```diff
@@ -0,0 +1,373 @@
+    compile_options.languageVersion = MTLLanguageVersion3_0;
+
+    for (size_t i = 0, e = iree_max(shader_library_count, shader_source_count); i < e; ++i) {
+      id<MTLLibrary> library = nil;
```
do these need to be released on failure? (line 308)

antiagainst replied:
Yes, done via `iree_hal_executable_destroy`.

benvanik commented on experimental/metal/registration/driver_module.c:
```diff
@@ -0,0 +1,87 @@
+#include "iree/base/tracing.h"
+
+IREE_FLAG(bool, metal_serial_command_dispatch, false,
+          "Run all commands in command encoder sequentially");
```
```suggestion
          "Serializes all commands within command buffers as if there were barriers between each.");
```

benvanik commented on experimental/metal/registration/driver_module.c:
```diff
@@ -0,0 +1,87 @@
+IREE_FLAG(bool, metal_serial_command_dispatch, false,
+          "Run all commands in command encoder sequentially");
+IREE_FLAG(bool, metal_command_buffer_retain_resources, false,
+          "Create command buffer to maintain strong references to resources");
```
```suggestion
          "Enables automatic Metal resource reference counting for diagnosing resource lifetime issues.");
```

benvanik commented on experimental/metal/registration/driver_module.c:
```diff
@@ -0,0 +1,87 @@
+IREE_FLAG(bool, metal_command_buffer_retain_resources, false,
+          "Create command buffer to maintain strong references to resources");
+IREE_FLAG(bool, metal_resource_hazard_tracking, false,
+          "Enable hazard tracking for resources");
```
```suggestion
          "Enables automatic Metal hazard tracking for diagnosing concurrency issues.");
```

benvanik commented on experimental/metal/metal_shared_event.h:
```diff
@@ -0,0 +1,49 @@
+// |listener| is used for dispatching notifications for async execution.
+//
+// |out_semaphore| must be released by the caller (see
+// |iree_hal_semaphore_release|).
```
```suggestion
// iree_hal_semaphore_release).
```

benvanik commented on experimental/metal/metal_shared_event.m:
```diff
@@ -0,0 +1,256 @@
+
+  IREE_TRACE_ZONE_END(z0);
+  if (timed_out) return iree_status_from_code(IREE_STATUS_DEADLINE_EXCEEDED);
+  return iree_ok_status();
```
if the semaphore failed this will need to return IREE_STATUS_ABORTED if any semaphore failed - you can set a did_fail flag in the notifyListener callback if v == UINT64_MAX and then return OK or ABORTED

benvanik commented on experimental/metal/metal_shared_event.m:
```diff
@@ -0,0 +1,256 @@
+
+  IREE_TRACE_ZONE_END(z0);
+  if (timed_out) return iree_status_from_code(IREE_STATUS_DEADLINE_EXCEEDED);
+  return iree_ok_status();
```
same as above with returning ABORTED - you can just did_fail = did_fail || v == UINT64_MAX above

benvanik commented on experimental/metal/pipeline_layout.h:
```diff
@@ -0,0 +1,100 @@
+// Creates a descriptor set layout for the given |bindings|.
+//
+// |out_descriptor_set_layout| must be released by the caller (see
+// |iree_hal_descriptor_set_layout_release|).
```
```suggestion
// iree_hal_descriptor_set_layout_release).
```

benvanik commented on experimental/metal/pipeline_layout.h:
```diff
@@ -0,0 +1,100 @@
+// |push_constant_count|.
+//
+// |out_pipeline_layout| must be released by the caller (see
+// |iree_hal_pipeline_layout_release|).
```
```suggestion
// iree_hal_pipeline_layout_release).
```

benvanik commented on experimental/metal/pipeline_layout.h:
```diff
@@ -0,0 +1,100 @@
+    iree_host_size_t push_constant_count,
+    iree_hal_pipeline_layout_t** out_pipeline_layout);
+
+// Retuns the descriptor set layout of the given |set| in
```
```suggestion
// Returns the descriptor set layout of the given |set| in
```

benvanik commented on experimental/metal/pipeline_layout.m:
```diff
@@ -0,0 +1,180 @@
+}
+
+iree_status_t iree_hal_metal_descriptor_set_layout_create(
+    iree_allocator_t host_allocator, iree_hal_descriptor_set_layout_flags_t flags,
```
by convention host_allocator immediately precedes the thing its used for - in creation functions that means `(....., host_allocator, out_resource)` so it should be moved before out_descriptor_set_layout

benvanik commented on experimental/metal/pipeline_layout.m:
```diff
@@ -0,0 +1,180 @@
+  IREE_TRACE_ZONE_END(z0);
+}
+
+iree_hal_descriptor_set_layout_binding_t* iree_hal_metal_descriptor_set_layout_binding(
```
return const

benvanik commented on experimental/metal/pipeline_layout.m:
```diff
@@ -0,0 +1,180 @@
+    iree_hal_descriptor_set_layout_t* base_descriptor_set_layout, uint32_t binding) {
+  iree_hal_metal_descriptor_set_layout_t* descriptor_set_layout =
+      iree_hal_metal_descriptor_set_layout_cast(base_descriptor_set_layout);
+  for (unsigned i = 0, e = descriptor_set_layout->binding_count; i < e; ++i) {
```
proper types, and no need for caching e as this is a basic read-only loop of a struct variable and it doesn't help

benvanik commented on experimental/metal/pipeline_layout.m:
```diff
@@ -0,0 +1,180 @@
+      iree_hal_metal_descriptor_set_layout_cast(base_descriptor_set_layout);
+  for (unsigned i = 0, e = descriptor_set_layout->binding_count; i < e; ++i) {
+    if (descriptor_set_layout->bindings[i].binding == binding)
+      return &descriptor_set_layout->bindings[i];
```
style: {} around multi-line ifs
(there being more macros in C make eliding them really dangerous)

benvanik commented on experimental/metal/staging_buffer.h:
```diff
@@ -0,0 +1,102 @@
+  // Current write offset of the device buffer.
+  uint32_t offset;
+
+  // The number of command buffers that are being recorded or executed on
```
as you note in your TODO above this is not going to work in real programs with any level of async - we'll want to get this fixed soon as it is a correctness issue. a suballocator is the easiest solution and maybe one day I can land that and we can just swap this out with that.

antiagainst replied:
https://github.com/openxla/iree/issues/14049 to track this.

antiagainst committed:
[metal] Use string view and byte span for compilation functions

antiagainst committed:
[metal] Order host_allocator and const-ify various query APIs

We want to have `host_allocator` directly ahead of the `out_`
parameters.

antiagainst committed:
[metal] Keep track of queue in buffer construction for macOS

For managed storage mode, we need to issue GPU commands to
explicitly make data visible to CPU. This would require
having a handle to a MTLCommandQueue. Instead of creating
a new one internally, give the queue to the buffer during
construction time.

antiagainst committed:
[metal] Check and return failure earlier in buffer allocation

This avoids the needs to create the buffer and clean it up later.

antiagainst committed:
[metal] Return early with IREE macro to flatten status check

This saves us one level of indent to make the code easier to
follow.

antiagainst committed:
[metal] Improve order in device creation

Change it to follow device struct allocation, initialization, and
subresource creation. This way we can handle destruction on failures
better.

antiagainst committed:
[metal] Cache a command buffer descriptor in device to deduplicate

antiagainst committed:
[metal] NFC: Make code in buffer fill less branchy

antiagainst committed:
[metal] Use the kernel layout to query push constant count

By this way, we can avoid tracking the current pipeline layout.

antiagainst committed:
[metal] Manage staging buffer refcount in command buffer lifetime

This avoids the potential refcount mismatches due to that we can
have command buffers created but never submitted and such.

antiagainst committed:
[metal] Use one resource set to handle queue execution resources

antiagainst committed:
[metal] Use the last command buffer for semaphore signaling

This is an performance improvement over the previous case where
we always create a new command buffer for semaphore signaling.

antiagainst committed:
[metal] Unify pipeline object creation in MTLLibrary and source paths

MTLFunction and MTLComputePipelineState creation are common between
these two paths; so we can have one function for them.

antiagainst committed:
[metal] Use pipeline layout to query set and binding count

This avoids mutating the `active_count` in the command buffer state.

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -337,50 +338,50 @@ iree_status_t iree_hal_metal_direct_command_buffer_create(
+  command_buffer->staging_buffer = staging_buffer;
+  command_buffer->host_allocator = host_allocator;
+  iree_status_t status = iree_hal_resource_set_allocate(block_pool, &command_buffer->resource_set);
+  iree_hal_metal_command_segment_list_reset(&command_buffer->segments);
```
you'll still want to guard all this below in is_ok(status) so that you don't set out_command_buffer/ref the staging buffer and instead release the command buffer before returning

antiagainst replied:
Ah, missed this part when refactoring.

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -337,50 +338,50 @@ iree_status_t iree_hal_metal_direct_command_buffer_create(
+
+  // Increase command buffer refcount in the shared staging buffer. We tie this to the command
+  // buffer's lifetime to avoid resource leak.
+  iree_hal_metal_staging_buffer_increase_refcount(staging_buffer);
```
you'll only want to stash the staging_buffer pointer on the command_buffer if status is ok - otherwise when you destroy below you may decrement the refcount even if you didn't increment it here due to the failure

antiagainst replied:
Good catch! Indeed. Fixed. Also fixed the missing destroy path in creation.

benvanik commented on experimental/metal/direct_command_buffer.m:
```diff
@@ -393,6 +394,10 @@ static void iree_hal_metal_command_buffer_destroy(iree_hal_command_buffer_t* bas
 
+  // Decrease command buffer refcount in the shared staging buffer, and potentially reclaim
+  // resources. We tie this to the command buffer's lifetime to avoid resource leak.
+  iree_hal_metal_staging_buffer_decrease_refcount(command_buffer->staging_buffer);
```
unless decrease_refcount handles NULL you'll want to NULL guard this in cases where the creation failed partway through

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -338,81 +337,61 @@ static iree_status_t iree_hal_metal_device_queue_execute(
   iree_hal_metal_device_t* device = iree_hal_metal_device_cast(base_device);
   IREE_TRACE_ZONE_BEGIN(z0);
 
+  iree_hal_resource_set_t* resource_set;
```
```suggestion
  iree_hal_resource_set_t* resource_set = NULL;
```
(for debugging sanity it's always good to 0 initialize in C)

benvanik commented on experimental/metal/metal_device.m:
```diff
@@ -338,81 +337,61 @@ static iree_status_t iree_hal_metal_device_queue_execute(
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_resource_set_allocate(&device->block_pool, &resource_set));
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
```
at this point if this function failed the resource_set would be leaked - you could do the normal status_is_ok chain or have the resource_set allocate and cleanup be in here and then call another function for actually doing the submission construction (like `resource_set_allocate, status = queue_submit(...); if (!iree_status_is_ok(status)) resource_set_free`)

antiagainst replied:
Ah, good catch. Makes sense. Fixed.

benvanik reviewed (changes requested):
omg, down to just a few minor things!

antiagainst committed:
[metal] Avoid resource set leak in queue execution

We need to check the status and free the resource set if any
error happens.

antiagainst committed:
[metal] Improve error handling in command buffer create/destroy

We need to check the return status of resource set allcoation
and properly destroy the command buffer if any error happens.
